### DISCUSSION TITLE
Synchronous request

### DISCUSSION THREAD

**Discussion by shial4** [stensoosaar/IBKit#9] [Ideas]
Some of the request could be done in a manner of synchronous way. 
We could allow for something like:

```
let requestID = client.nextRequestID
let result: IBContractSearchResult = try await client.searchSymbols(requestID, nameOrSymbol: "MES")
```

or 

```
let requestID = client.nextRequestID
let bars = try client.requestPriceHistory(
            requestID,
            contract: contract,
            barSize: size,
            barSource: source,
            lookback: duration 
)
```

where we pair up request with response delivering synch experience.
Obviously not every request would work with this approach but some would.
Something like [ib_insync](https://github.com/erdewit/ib_insync) was delivering.

**Comment by stensoosaar** [comment]
In near future there will be two options to make api calls:
```
let request = IBRealTimeBarRequest(requestID: requestID, contract: contract, source: .midpoint, extendedTrading: true)
try client.send(request)
or using convenience method
try client.subscribeRealTimeBar(requestID, contract: contract, barSource: .midPoint, extendedTrading: true)
```
As there are many pacing and concurrent request limitations, it could be useful to be able to keep track of your filled and unfilled tasks (if you have large trading universe)

What concerns pairing up request and response, i guess it depends about the end user user preferences. if you have one contract, one timeframe trading strategy, you probably don't want to waste any time as it doesn't much matter. On the other hand, more complex strategies will need custom object mapping (using request info).

I have played with following idea:
```
public func priceHistoryPublisher(_ interval: DateInterval, resolution: IBBarSize, contract: IBContract, extendedSession: Bool = false) throws -> AnyPublisher<TKPriceHistory,IBResponseError>{
		
		do {
			let requestID = api.nextRequestID
			let source: IBBarSource = [.cfd, .forex].contains{$0 == contract.securitiesType} ? .midpoint : .trades
			let request = IBPriceHistoryRequest(requestID: requestID, contract: contract, size: resolution, source: source, lookback: interval, useRTH: !extendedSession)
			try api.send(request: request)
			
			
			return api.events
				.setFailureType(to: IBResponseError.self)
				.compactMap { $0 as? IBRequestResponse }
				.filter { $0.requestID == requestID }
				.tryMap { response -> IBPriceHistory in
					if let sampleMessage = response as? IBPriceHistory {
						return sampleMessage.convert()
					} else if let error = response as? IBResponseError {
						throw error
					} else {
						throw IBResponseError(requestID: requestID, code: -1, message: "Unknown response for index \(requestID)}")
					}
				}
				.mapError { $0 as! IBResponseError }
				.eraseToAnyPublisher()
		} catch {
			throw error
		} 
```

and using it to map response object to my own frameworks one ( I noticed you were aleo using your own custom bars). 

Maybe some custom operators or publishers to make the writing your custom datasource publishers a bit easier?
